src/static/js/caretPosition.js

```diff
@@ -0,0 +1,236 @@
+// One rep.line(div) can be broken in more than one line in the browser.
+// This function is useful to get the caret position of the line as
+// is represented by the browser
+exports.getPosition = function ()
+{
+  var rect, line;
+  var editor = $('#innerdocbody')[0];
+  var range = getSelectionRange();
+  var isSelectionInsideTheEditor = range && $(range.endContainer).closest('body')[0].id === 'innerdocbody';
+
+  if(isSelectionInsideTheEditor){
+    // when we have the caret in an empty line, e.g. a line with only a <br>,
+    // getBoundingClientRect() returns all dimensions value as 0
+    var selectionIsInTheBeginningOfLine = range.endOffset > 0;
+    if (selectionIsInTheBeginningOfLine) {
+      var clonedRange = range.cloneRange();
+
+      // we set the selection start and end to avoid error when user selects a text bigger than
+      // the viewport height and uses the arrow keys to expand the selection. In this particular
+      // case is necessary to know where the selections ends because both edges of the selection
+      // is out of the viewport but we only use the end of it to calculate if it needs to scroll
+      clonedRange.setStart(range.endContainer, range.endOffset);
+      clonedRange.setEnd(range.endContainer, range.endOffset);
+
+      rect = clonedRange.getBoundingClientRect();
+      line = {
+        bottom: rect.bottom,
+        height: rect.height,
+        top: rect.top
+      }
+      clonedRange.detach();
+    }
+    // in this case, we can't get the dimension of the element where the caret is
+    if(!rect || rect.height === 0){
+      clonedRange = range.cloneRange();
+      clonedRange.setStart(range.endContainer, range.endOffset);
+      clonedRange.setEnd(range.endContainer, range.endOffset);
+
+      // as we can't get the element height, we create a text node to get the dimensions
+      // on the position
+      var shadowCaret = $(document.createTextNode("|"));
+      clonedRange.insertNode(shadowCaret[0]);
+      clonedRange.selectNode(shadowCaret[0]);
+
+      rect = clonedRange.getBoundingClientRect();
+      line = {
+        bottom: rect.bottom,
+        height: rect.height,
+        top: rect.top
+      }
+      shadowCaret.remove();
+      clonedRange.detach();
+    }
+  }
+  return line;
+}
+
+exports.getPositionOfRepLineAtOffset = function (node, offset) {
+  // it is not a text node, so we cannot make a selection
+  if (node.tagName === 'BR' || node.tagName === 'EMPTY') {
+    return getPositionOfElementOrSelection(node);
+  }
+
+  if (node.length === 0){
+    while (node.length === 0 && node.nextSibling) {
+      node = node.nextSibling;
+    }
+  }
+
+  var newRange = new Range();
+  newRange.setStart(node, offset);
+  newRange.setEnd(node, offset);
```

Comment:
Is it possible that the offset would be inside a more internal node on the DOM? For example:

```
<span id='outer'>
  123
  <span id='inner'>
    456
  </span>
  789
</span>
```

`node` = `#outer`
`offset` = `5`

What would happen in this case?

Reply:
You'd get an error. This function is called twice
`getPositionOfRepLineAtOffset(firstRootChildNode, 1)`
`getPositionOfRepLineAtOffset(lastRootChildNode.node, lastRootChildNode.length)`
In both cases is not possible to get the inner node as the example.